### PR TITLE
Pull image before tests execution.

### PR DESCRIPTION
**Pull request by phracek**
Fail in case of error.

Required-by: s2i-python-container#449

Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

### CODE REVIEW

**Comment by phracek** [comment]
[test]

File: test-lib.sh [hunk 1]
```diff
@@ -52,6 +52,42 @@ function ct_enable_cleanup() {
...
+
+  # Try pulling the image to see if it is accessible
+  # WORKAROUND: Since Fedora registry sometimes fails randomly, let's try it more times
+  while ! docker pull "$image_name"; do
+    ((loop++)) || :
+    echo "Pulling image $image_name failed."
+    if [ "$loop" -gt "$loops" ]; then
+      echo "It happened $loops times. Giving up."
+      echo "!!! ERROR with pulling image $image_name !!!!"
+      return 1
```
Comment (phracek) [inline test-lib.sh:83]: This return value can be configurable and by default `return 1`. Otherwise, `exit 1` in case of checking image availability before tests execution.

The function itself can be used in all `container-common-scripts`.

File: test-lib.sh [hunk 2]
```diff
@@ -765,20 +801,6 @@ EOF
...
-ct_check_image_availability() {
-  local public_image_name=$1;
-
-  # Try pulling the image to see if it is accessible
-  if ! docker pull "$public_image_name" &>/dev/null; then
-    echo "$public_image_name could not be downloaded via 'docker'"
-    return 1
-  fi
-}
-
```
Comment (pkubatrh) [inline test-lib.sh:781]: Can we keep the function and just call `ct_pull_image` inside? I would rather not remove the function completely in case it is used in some of the repositories

File: test-lib.sh [hunk 3]
```diff
@@ -52,6 +52,42 @@ function ct_enable_cleanup() {
...
+# Argument: image_name - string containing the public name of the image to pull
+# Function returns either 0 in case of pull was successful
+# Or the test suite exit with 1 in case of pull error
+function ct_pull_image() {
+  local image_name="$1"
+  local loops=10
+  local loop=0
+
+  # Let's try to pull image.
+  echo "-> Pulling image $image_name before tests execution."
```
Comment (pkubatrh) [inline test-lib.sh:67]: I would remove the `before tests execution.` part from the output. It might confuse the reader in case the function is actually not used before executing tests.

File: test-lib.sh [hunk 4]
```diff
@@ -52,6 +52,42 @@ function ct_enable_cleanup() {
...
 
+# ct_pull_image
+# -------------
+# Function pull an image before tests execution
+# Argument: image_name - string containing the public name of the image to pull
+# Function returns either 0 in case of pull was successful
+# Or the test suite exit with 1 in case of pull error
+function ct_pull_image() {
+  local image_name="$1"
+  local loops=10
```
Comment (pkubatrh) [inline test-lib.sh:63]: How about making it possible to set the amount of loops via function argument

File: test-lib-openshift.sh [hunk 5]
```diff
@@ -779,7 +779,6 @@ function ct_os_test_template_app_func() {
       for i_t in ${other_images} ; do
         echo "${i_t}"
         IFS='|' read -ra image_tag_a <<< "${i_t}"
-        docker pull "${image_tag_a[0]}"
```
Comment (pkubatrh) [inline test-lib-openshift.sh:782]: Just wondering if this change is expected as it does not seem to be related to the rest of the changes

File: test-lib.sh (hunk 2 above)
Comment (phracek) [reply test-lib.sh:781]: Sure. I will do it.

File: test-lib.sh (hunk 4 above)
Comment (phracek) [reply test-lib.sh:63]: I will do it so, that If the parameter is not specified, then used default.

File: test-lib-openshift.sh (hunk 5 above)
Comment (phracek) [reply test-lib-openshift.sh:782]: Good point. I wanted to remove docker pull, because of `docker pull` or especially pulling image has to be done at the beginning like here: https://github.com/sclorg/s2i-python-container/pull/449/files#diff-0b78990f6549655ad2c4497f3e0b771c74d42fc70837c7e1222728ae2cd566a4R20

But maybe we can add there a check if an image does not exist, then failed and say user to add ct_pull_image at the beginning of tests.

Comment (phracek) [reply test-lib-openshift.sh:782]: See me proposal if it make sense to you.

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
Failed because of docker hub rate limit.

[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
@pkubatrh Can you please look at this pull request once more? It looks like it help us with testing on OpenShift 3. See error here: http://artifacts.dev.testing-farm.io/54a6352b-9b53-4bdd-9f1f-2200aef34e89/

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
Rebased

[test]

**Comment by phracek** [comment]
PR is blocked by https://github.com/sclorg/container-common-scripts/pull/203

**Comment by phracek** [comment]
Rebased

[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

File: test-lib-openshift.sh (hunk 5 above)
Comment (pkubatrh) [reply test-lib-openshift.sh:828]: Makes sense, thanks!

File: test-lib.sh [hunk 6]
```diff
@@ -52,6 +52,47 @@ function ct_enable_cleanup() {
   trap ct_cleanup EXIT SIGINT
 }
 
+# ct_pull_image
+# -------------
+# Function pull an image before tests execution
+# Argument: image_name - string containing the public name of the image to pull
```
Comment (pkubatrh) [inline test-lib.sh:58]: Can you add docstrings for the new arguments as well?

File: test-lib.sh [hunk 7]
```diff
@@ -52,6 +52,47 @@ function ct_enable_cleanup() {
...
+  local exit=${1:-"false"}
+  local loop=0
+
+  # Let's try to pull image.
+  echo "-> Pulling image $image_name ..."
+  # Sometimes in Fedora case it fails with HTTP 50X
+  # Check if the image is available locally and try to pull it if it is not
+  if [[ "$(docker images -q "$image_name" 2>/dev/null)" != "" ]]; then
+    echo "The image $image_name is already pulled."
+    if [[ x"$exit" == "xfalse" ]]; then
```
Comment (pkubatrh) [inline test-lib.sh:73]: I do not see any calls to `ct_pull_image` with `exit` set. Do you already have some use case in mind for future calls where this can be useful to have?

File: test-lib.sh [hunk 8]
```diff
@@ -52,6 +52,47 @@ function ct_enable_cleanup() {
...
+    fi
+  fi
+
+  # Try pulling the image to see if it is accessible
+  # WORKAROUND: Since Fedora registry sometimes fails randomly, let's try it more times
+  while ! docker pull "$image_name"; do
+    ((loop++)) || :
+    echo "Pulling image $image_name failed."
+    if [ "$loop" -gt "$loops" ]; then
+      echo "It happened $loops times. Giving up."
```
Comment (pkubatrh) [inline test-lib.sh:86]: I owuld suggest a bit more verbose error message here.

```suggestion
      echo "Pulling of image $image_name failed $loops times in a row. Giving up."
```

File: test-lib.sh (hunk 7 above)
Comment (phracek) [reply test-lib.sh:73]: This is a great catch. I have a mistake in the script. I will update the function.

File: test-lib-openshift.sh [hunk 9]
```diff
@@ -825,7 +825,13 @@ function ct_os_test_template_app_func() {
...
+        if [[ "$(docker images -q "$image_name" 2>/dev/null)" != "" ]]; then
```
Comment (phracek) [inline test-lib-openshift.sh:828]: In case the image is already pulled, then this condition is wrong.

It should be `if [[ "$(docker images -q "$image_name" 2>/dev/null)" == "" ]]; then`

**Comment by phracek** [comment]
[test]

**Comment by pkubatrh** [comment]
Seems like the failing tests are because one of the images used in tests cannot be built since a Dockerfile is not available: https://github.com/sclorg/container-common-scripts/tree/master/tests/failures/check/v0

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

File: test-lib.sh [hunk 10]
```diff
@@ -185,7 +228,7 @@ function ct_assert_container_creation_fails() {
 function ct_create_container() {
   local cid_file="$CID_FILE_DIR/$1" ; shift
   # create container with a cidfile in a directory for cleanup
-  # shellcheck disable=SC2086
+  # shellcheck disable=SC2086,SC2153
```
[the remark this answers is not in the document]
Comment (phracek) [reply test-lib.sh:231]: The reason is `^---------^ SC2153: Possible misspelling: IMAGE_NAME may not be assigned, but image_name is.`

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]
